scalyr_agent/util.py

```diff
@@ -1337,3 +1341,265 @@ def close(self):
+    """
+
+    INCREASE_STRATEGY_MULTIPLY = 'multiply'
+    INCREASE_STRATEGY_RESET_THEN_MULTIPLY = 'reset_then_multiply'
+
+    def __init__(self, num_agents, initial_cluster_rate, upper_cluster_rate, lower_cluster_rate,
+                 consecutive_success_threshold, increase_strategy, increase_factor=2.0, backoff_factor=2.0,
+                 max_concurrency=1, fake_clock=None):
+        """
+        @param num_agents: Number of agents in the cluster
+        @param initial_cluster_rate: initial cluster rate (requests/sec)
+        @param upper_cluster_rate: upper bound on cluster rate (requests/sec)
+        @param lower_cluster_rate: lower bound on cluster rate (requests/sec)
+        @param consecutive_success_threshold: number of consecutive successes to trigger a rate increase
+        @param increase_strategy: strategy for increasing the rate (multiply by increase_factor or reset to initial)
+        @param increase_factor: multiplicative factor for increasing rate
+        @param backoff_factor: divisive factor for decreasing rate
+        @param max_concurrency: max number of tokens (for limiting concurrent requests) that can be acquired
+
+        @type num_agents: int
+        @type initial_cluster_rate: float
+        @type upper_cluster_rate: float
+        @type lower_cluster_rate: float
+        @type consecutive_success_threshold: int
+        @type increase_strategy: str
+        @type increase_factor: float
+        @type backoff_factor: float
+        @type max_concurrency: int
+        """
+        self._num_agents = num_agents
+        self._initial_cluster_rate = initial_cluster_rate
+        self._upper_cluster_rate = upper_cluster_rate
+        self._lower_cluster_rate = lower_cluster_rate
+        self._consecutive_success_threshold = consecutive_success_threshold
+        self._increase_strategy = increase_strategy
+        self._increase_factor = increase_factor
+        self._backoff_factor = backoff_factor
+
+        self._current_cluster_rate = self._initial_cluster_rate
+        self._consecutive_successes = 0
+        self._cluster_rate_lock = threading.Lock()
+
+        self._max_concurrency = max_concurrency
+
+        # A queue of tokens
+        self._ripe_time = None
+        self._token_queue = deque()
+        # Condition variable to synchronize access to token queue
+        self._token_queue_cv = threading.Condition()
+
+        # fake clock for testing
+        self._fake_clock = fake_clock
+        self._test_mode_lock = threading.Lock()
+
+        # Validate input
+        strategies = [self.INCREASE_STRATEGY_MULTIPLY, self.INCREASE_STRATEGY_RESET_THEN_MULTIPLY]
+        if increase_strategy not in strategies:
+            raise ValueError('Increase strategy must be one of %s' % strategies)
+
+        if self._max_concurrency < 1:
+            raise ValueError('max_concurrency must be greater than 1')
+
+        if self._consecutive_success_threshold <= 1:
```

Comment:
Why not allow the `consecutive_success_threshold` to be 1?

Reply:
good catch.